src/ModelContextProtocol.Core/Protocol/JsonRpcMessage.cs

```diff
@@ -75,6 +78,48 @@ private protected JsonRpcMessage()
     [EditorBrowsable(EditorBrowsableState.Never)]
     public sealed class Converter : JsonConverter<JsonRpcMessage>
     {
+        /// <summary>
+        /// The union to deserialize.
+        /// </summary>
+        internal struct Union
```

Comment:
@Scooletz I would recommend marking the `Union` struct as private and hand-craft a static `Parse(ref Utf8JsonReader reader)` that populates an instance of the type. This should further improve performance while also minimizing the amount of source generated code (aka there wouldn't exist any serialization code for the type).